lib/pbench/server/api/resources/query_apis/__init__.py

```diff
@@ -77,45 +106,145 @@ def __init__(
+            AssemblyError: Reports an unsupported access combination we expect
+                to be prohibited by upstream authorization checks. If such a
+                case reaches here, query assembly will be aborted.
 
         Returns:
-            An assembled Elasticsearch query term
+            An assembled Elasticsearch "query" mode that includes the necessary
+            user/access terms.
         """
-        user = json.get("user")
+        user = parameters.get("user")
+        access = parameters.get("access")
+        authorized_user: User = Auth.token_auth.current_user()
+        authorized_id = str(authorized_user.id) if authorized_user else None
+        is_admin = authorized_user.is_admin() if authorized_user else False
+
+        filter = terms.copy()
+        self.logger.debug(
+            "QUERY auth ID {}, user {!r}, access {!r}, admin {}",
+            authorized_id,
+            user,
+            access,
+            is_admin,
+        )
+
+        # If a user is specified, assume we'll be selecting for that user
         if user:
-            term = {"authorization.owner": user}
-        elif json.get("access") == Dataset.PRIVATE_ACCESS:
-            raise UnsupportedAccessMode(user, Dataset.PRIVATE_ACCESS)
+            filter.append({"term": {"authorization.owner": user}})
+
+        # IF we're looking at a user we're not authorized to see (either the
+        # call is unauthenticated, or authenticated as a non-ADMIN user trying
+        # to reference a different user), then we're only allowed to see PUBLIC
+        # data; private data requests aren't allowed, and we expect these cases
+        # to fail with appropriate permission errors (401 or 403) before we
+        # reach this code.
+        #
+        # ELSE IF we're looking for a specific access category, add the query
+        # term.
+        #
+        # ELSE IF this is a {} query from a non-ADMIN user, add a "disjunction"
+        # (OR) covering all the user's own data plus all public data.
+        #
+        # ELSE this must be an admin user looking for data owned by a specific
+        # user (there's no access parameter, and possibly no user parameter),
+        # and we'll pass through the query either with the specified user
+        # constraint or with no constraint at all.
+        if not authorized_user or (user and user != authorized_id and not is_admin):
+            if access == Dataset.PRIVATE_ACCESS:
+                raise AssemblyError(
+                    f"Internal error: can't generate query for user {user}, access {access}"
+                )
+            filter.append({"term": {"authorization.access": Dataset.PUBLIC_ACCESS}})
+            self.logger.debug("QUERY: not self public: {}", filter)
+        elif access:
+            filter.append({"term": {"authorization.access": access}})
+            if not user and not is_admin:
+                filter.append({"term": {"authorization.owner": authorized_id}})
```

Comment:
I think you may be partly correct; I'll have to investigate this further. I thought I had a test case: either it's missing or it's broken. Yes, an authorized user asking for `{"access": "public"}` should see everyone's public data.

However, an authorized user (or admin) asking for another user's public data (or not constraining user) *does* need an access filter, because Elasticsearch will happily return private data without it.

Reply:
If control gets to 226, then we'll add the `"access"` filter that we need.  But, we only need to add `{"authorization.owner": authorized_id}` if we're requesting `{"access": "private"}` (f we didn't add it at line 199) and if we're not admin.